Correct cp pipeline and rerun

jenna-tomkinson opened the pull request:
@gwaybio 

Here is the partial PR for NF1. I was able to rerun the pipelines with the correction for analyzing the whole image features (the orig_nuclei issue was handled in previous PR). I was able to get the All CellProfiler pipelines for Plate 1 and 2 ran for everything (except heatmaps for Plate 2).

The common pattern I am seeing is that very pipeline is giving different results.

gwaybio commented:
@jenna-tomkinson - I pulled this branch and I'm currently working through this PR. It's tough to figure out what's going on! There are a ton of files, and it's unclear to me how they've been generated.

We had talked about creating a separate benchmarking repository. I'm wondering if it's a better option to move the multi-pipeline benchmarking to a separate repo and close this PR in this repo. This repo can focus on the core CP pipeline we've been using. What do you think? I hesitate to merge this not being able to track provenance of figures.

jenna-tomkinson commented:
@gwaybio They are all generated using the pipelines originally made. I agree there are a ton of files.

What I think would be best is that I will make a slideshow with all the figures for the core CP pipeline so that we can look them over. As for the repos, I am thinking we might make two new repos at this point because this repo wouldn't be good to have as the main CP pipeline since there is a lot in it that is not apart of the CP pipeline (mainly cause of DP). This would be good to have a meeting over.

gwaybio commented:
Sounds good. Let's discuss next week.

gwaybio commented:
@jenna-tomkinson - it seems the plates are more similar than maybe we initially thought - see recent commits

![plate1_vs_2_coefficients](https://user-images.githubusercontent.com/7340421/219663834-77410379-aa93-409a-9b83-4f5b08e7f2f9.png)

jenna-tomkinson commented:
@gwaybio 

I am planning on creating a new commit in this PR with the following changes (essentially refactoring):

1. Edit CellProfiler analysis pipelines:

- Remove the extra `MeasureGranularity` module (learned that it does both images and objects)
- Remove and change `MeasureObjectNeightbors` to measure adjacent neighbors (was set to a specific distance)

2. Change structure of running the CellProfiler pipelines to using a notebook and the command line (similar to what is done in the CFReT project) and rerun all methods with the updated pipelines.
3. Update the extraction of single cells to have a utility function that will call the SingleCell class, merge, normalize, and feature selection so that all methods can be performed in one notebook.
4. Remove all extra figures that were created in the analysis module except for those for the all CellProfiler method and those created by @gwaybio

jenna-tomkinson commented:
@gwaybio 

This PR has been refactored and is ready for review!

To make this easier, I recommend reviewing per commit as each commit is pretty small unless it deletes files. See below for a detailed list of each commit. 

1. [daf19af](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/daf19afbe2331db171e5ab87782dde1721c8a47c): Create functions to run each method through CellProfiler and rename the SQLite file to reflect the method used (e.g., CellProfiler IC + Cellpose segmentation = `cellprofiler_cellpose`). As well, there are two notebooks, one for each plate, that uses these functions.
2. [882612d](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/882612da48a7100a57af2c384eded69e8ead0b2d): In this commit, I removed all old SQLite files from previous run and added the new SQLite files that were acquired using the new method implemented in the previous commit.
3. [8a2ef5b](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/8a2ef5b3886338be8919c24506e80c4a6754b75a): The CellProfiler pipelines using CP segmentation and Cellpose segmentation have been updated to reflect the changes I stated in the previous comment.
**Note:** Ignore commits [899ebbf](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/899ebbf0e7302157f629693e386be5c7fb9c0a95) and [b212cf5](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/b212cf5592f6bf48734f81d64792feaa65a0b345) as one contains an version of a file that is updated in another commit and the other only deletes files.
4. [baa41bd](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/baa41bd32d5fd735778711b36a8ab589c7393c9d): This commit contains major README changes to reflect the three new notebooks that run all CP and DP methods for each plate. There is a file with functions to run Pycytominer (essentially Cytosnake 🤣) which makes it easier to run multiple methods in one notebook.
5. [6f5295e](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/6f5295e05cdef37f9da3b4a3f3f2fcbdf00645d5): Add all new `csv.gz` files with extracted single cells from each method.
6. [893609f](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/893609f2dc8ff5402e635bb8d77a716b716ea9f9): This last commit goes into the analysis module and remove all figures from running the notebooks with any other method other than the `All CellProfiler` method or DeepProfiler with plate 1. This should revert all notebooks back to how they are currently in the repo.

I apologize that this needed such a refactor, but I think this will be easier to work with and is much cleaner.

roshankern commented on 4_processing_features/extract_single_cells.sh:
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+# CellProfiler methods
```
Consider rewriting this comment and the other in this file to clarify what these nbconvert commands do.

jenna-tomkinson replied:
Fixed in commit [e2512de](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/e2512de1d4f51232a14dc02f74361e09342550a0)

roshankern commented on 4_processing_features/extraction_utils.py:
```diff
@@ -0,0 +1,159 @@
+import pathlib
```
Consider adding docstring to the top of this file to clarify what this file is for.

jenna-tomkinson replied:
Makes sense to add that documentation, added now!

roshankern commented on 4_processing_features/scripts/plate1_extract_sc_cp.py:
```diff
@@ -0,0 +1,226 @@
+
+
+# Set file and directory constants
+cp_dir = "../CellProfiler_pipelines"
```
Consider using `pathlib.Path` for these paths if possible (as per Way Lab conventions)

roshankern replied:
^This comment applies to the rest of the paths in this file as well!

jenna-tomkinson replied:
Good point! I just added these to all paths but it will not work for the paths to SQLite files as it needs to be a string for the package to read the file. I have fixed all of the paths in the 4_processing_features module in commit [370b0da](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/370b0da035ee3d9f6fc5f7cbb65df301e46929b8).

roshankern commented on 4_processing_features/scripts/plate1_extract_sc_cp.py:
```diff
@@ -0,0 +1,226 @@
+# In[11]:
+
+
+pd.crosstab(data_df.Metadata_genotype, data_df.Metadata_Well)
```
Consider adding a comment to explain what this function does.

roshankern replied:
^Given your use of `crosstab` throughout this file, I think it would be enough to add a comment here and nowhere else in this file.

jenna-tomkinson replied:
Added in commit [e2512de](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/e2512de1d4f51232a14dc02f74361e09342550a0)

roshankern commented on 5_analyze_data/README.md:
````diff
@@ -0,0 +1,52 @@
+```sh
+jupyter lab
+```
+This will open Jupyter Lab into your browser. 
````
```suggestion

This will open Jupyter Lab into your browser. 
```
Formatting

jenna-tomkinson replied:
Fixed in [f66e49a](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/f66e49a88b1ca6f0b20d5be66096be358ed24d8f)

roshankern commented on 5_analyze_data/notebooks/KS-test_analysis/kstest_utils.py:
```diff
@@ -0,0 +1,75 @@
+from scipy.stats import ks_2samp
```
Consider adding a docstring to clarify what this file is for.

jenna-tomkinson replied:
Added in [f66e49a](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/f66e49a88b1ca6f0b20d5be66096be358ed24d8f)

roshankern commented on 5_analyze_data/notebooks/KS-test_analysis/kstest_utils.py:
```diff
@@ -0,0 +1,75 @@
+import pathlib
+import pandas as pd
+
+def nf1_ks_test_two_sample(normalized_data: pd.DataFrame):
```
```suggestion
def nf1_ks_test_two_sample(normalized_data: pd.DataFrame) -> pd.DataFrame:
```

jenna-tomkinson replied:
Added in [f66e49a](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/f66e49a88b1ca6f0b20d5be66096be358ed24d8f)

roshankern commented on 5_analyze_data/notebooks/KS-test_analysis/kstest_utils.py:
```diff
@@ -0,0 +1,75 @@
+    feature_results: pd.DataFrame,
+    feature_names: list,
+    save_path: pathlib.Path = None,
+):
```
```suggestion
) -> pd.DataFrame:
```

I think this syntax is correct, please make sure! We want to hint return types as well

jenna-tomkinson replied:
Yes this is what it returns and I have added it to both functions in [f66e49a](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/f66e49a88b1ca6f0b20d5be66096be358ed24d8f)

roshankern commented on 5_analyze_data/notebooks/compare_plates/nbconverted/compare_plate1and2.r:
```diff
@@ -0,0 +1,95 @@
+    paste0(colnames(plate_2_df)[seq(2, ncol(plate_2_df))], "_plate2")
+)
+
+plate_df <- plate_1_df %>%
```
Consider adding a comment to explain what this chunk of code does

jenna-tomkinson replied:
Since @gwaybio created this code for this PR, I am not 100% sure what it actually does (still learning R). I will refrain from adding a comment here at this time, but I agree that it would be good to have it here.

roshankern commented on 5_analyze_data/visualize_env.yml:
```diff
@@ -24,4 +24,5 @@ dependencies:
   - conda-forge::r-reshape2
   - conda-forge::r-cowplot
   - conda-forge::r-patchwork
+  - conda-forge::r-pwr
```
Consider specifying package versions (always)!

jenna-tomkinson replied:
As well, since @gwaybio originally made the PR for the `5_analyze_data` module and is more familiar with the packages, I refrain from adding any exact specific package versions. I am trying to get into the adding them for other environment now, so maybe I can `conda list` and add the versions but it might not be the ones that Greg has when he originally made it.

roshankern commented on CellProfiler_pipelines/Pipelines/NF1_analysis_cp_plugin.cppipe:
```diff
@@ -65,7 +65,7 @@ Groups:[module_num:4|svn_version:'Unknown'|variable_revision_number:2|show_windo
     Metadata category:None
 
-RunCellpose:[module_num:5|svn_version:'Unknown'|variable_revision_number:3|show_window:False|notes:[]|batch_state:array([], dtype=uint8)|enabled:True|wants_pause:False]
+RunCellpose:[module_num:5|svn_version:'Unknown'|variable_revision_number:3|show_window:False|notes:['The default is 15 for minimum size that is used in Cellpose function, but I have it turned off. Can be defaulted back if needed.']|batch_state:array([], dtype=uint8)|enabled:True|wants_pause:False]
```
Is it worth documenting somewhere why you decided to turn it off?

jenna-tomkinson replied:
I am going to hold off on that since this was something I figured out was an unnecessary change in IDR_stream and I already have it set to turn off here since I assumed it is turned off in the Cellpose GUI. This comment was mainly for me, and I don't know how to explain why I turned it off expect for I thought that was how the GUI did it.

roshankern commented on CellProfiler_pipelines/Pipelines/NF1_analysis_cp_plugin.cppipe:
```diff
@@ -88,7 +88,7 @@ RunCellpose:[module_num:5|svn_version:'Unknown'|variable_revision_number:3|show_
     Remove objects that are touching the edge?:Yes
 
-RunCellpose:[module_num:6|svn_version:'Unknown'|variable_revision_number:3|show_window:False|notes:[]|batch_state:array([], dtype=uint8)|enabled:True|wants_pause:False]
+RunCellpose:[module_num:6|svn_version:'Unknown'|variable_revision_number:3|show_window:False|notes:['Same as previous module.']|batch_state:array([], dtype=uint8)|enabled:True|wants_pause:False]
```
I think it would be clearer to copy the description from the last file here in addition to the (same as previous module).

jenna-tomkinson replied:
Agreed, I have changed it and added the why to turning it too (e.g. I thought it should be turned off) in commit [b1f5958](https://github.com/WayScience/NF1_SchwannCell_data/pull/42/commits/b1f59587b6e32321559f1514c6710f0363369155)

roshankern commented on CellProfiler_pipelines/Pipelines/NF1_analysis_cp_plugin.cppipe:
```diff
@@ -2,7 +2,7 @@ CellProfiler Pipeline: http://www.cellprofiler.org
 Version:5
```
How is this file used in conjunction with `NF1_analysis.cpipe`? They seem to have nearly identical changes in this PR.

jenna-tomkinson replied:
Since in this repo we have multiple different image-analysis pipelines, we have to have two different analysis cppipe files to run different segmentation method. Essentially, the two analysis pipelines need to be the extract same except for segmentation to be able to perform benchmarking in the future.

roshankern commented on CellProfiler_pipelines/cputils.py:
```diff
@@ -0,0 +1,68 @@
+        pathlib.Path(sqlite_file_path).rename(pathlib.Path(new_file_name))
+        print(f"The file is renamed to {pathlib.Path(new_file_name).name}!")
+
+    except FileNotFoundError as e:
```
It might be better to `raise` an exception here instead (or in addition to) printing something out. This way the code won't keep running, unless that is your intention!

jenna-tomkinson replied:
I am not 100% familiar with raise versus except, so in this case I don't really have an intention. I just know that this method of except will raise an error if there is no file out of all the files named `NF1_data.sqlite` so that means there is a problem. I think this will work just fine for this case as to me if I change it, it wouldn't improve the time it take to run.

I will look into `raise` in the future though since it might be best to use that more than try/excepts so thank you for bringing it up!

roshankern reviewed (commented):
118 files to view! Consider using the `gitignore` file to ignore binary files (like those in `data/`) until the PR is approved. This way, the reviewer does not need to approve all of the added binary files or PNGs.

In cases where you need to process multiple plates/feature types (CP&DP), I think it is better to use for loops (iterate through plates or feature types). There isn't anything to change for this PR I just wanted to get this on your mind for future reference!

After looking at my comments this looks good to merge to me! Great job with the documentation of your code here. Of course, try to make smaller PRs in the future :)

jenna-tomkinson commented:
@roshankern Thank you for the review! The next PRs will be short I promise (gotta make mistakes to be able to learn :smile:). 

@gwaybio I will now merge this PR and it should be good to go for Cameron to use and test NF1 analysis with.
